### DISCUSSION TITLE
question about rust

### DISCUSSION THREAD

**Discussion by FayneBupt** [databendlabs/databend#4225] [Q&A] [answered]
I know that the Firebolt project is also developed in Clickhouse, but they are still written in C++, which saves a lot of coding time. Why doesn't databend do this and rewrite it with trust？
At present, I haven't found the open source code to transform Clickhouse into MPP architecture，which solving the problem of distributed join and separating computer and storage.

**Comment by bohutang** [comment, marked as answer]
Good question.
We evaluated doing Storage and Compute Disaggregated on ClickHouse and found it's hard([There is a post about it](https://bohutang.me/2020/09/18/clickhouse-and-friends-compute-storage/)).

On the fact that all current data warehouses code-base suffer from this problem, they are not designed for cloud architectures and are difficult to maintain and grow sustainably.

So, we writing one from scratch in Rust:
* Storage and Compute Disaggregated from day-1
* Vectorized Execution Engine and Push&Pull-Based Processor Model
* Elastic, Elastic, Elastic

Current amount of code is about 131293 lines(From tokei).

**Reply by FayneBupt** [reply]
https://zhuanlan.zhihu.com/p/443209576        How do you evaluate this scheme, integrate Doris's MPP architecture into Clickhouse, and reuse Clickhouse's excellent vectorization code, storage code，Push&Pull-Based Processor Model

**Reply by bohutang** [reply]
This is a good solution so far, but the complexity and maintenance costs are still quite high :)

**Reply by FayneBupt** [reply]
n fact, we are doing similar things....
However, compared with Clickhouse and Doris, a big reason for their success is the support of the internal use of the manufacturers behind them;
Therefore, compared with Firebolt, I still admire you very much. Only those who have experienced it know how difficult it is to build these from scratch..

**Comment by lichuang** [comment]
> Why rewrite it with rust？

Hmm, before I join databend, i am a C\C++ programmer for more than 10 years, after 2 month writing rust code, i am in love with this language.In my opition, Rust fix the core problem in C\C++: the runtime memory bug, such as wild pointer access,memory leaks,buffer overflow, etc.

I have write an article about the experience of rust(in Chinese): [周刊（第7期）：一个C系程序员的Rust初体验 - codedump的网络日志](https://www.codedump.info/post/20220227-weekly-7/)
